src/man/man3/sam_import.3
.\" $Revision: 1.17 $
.ds ]W Sun Microsystems
.\" SAM-QFS_notice_begin
.\"
.\" CDDL HEADER START
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
.\" or https://illumos.org/license/CDDL.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at pkg/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" CDDL HEADER END
.\"
.\" Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
.\" Use is subject to license terms.
.\"
.\" SAM-QFS_notice_end
.TH sam_import 3 "07 Jan 2009"
.SH NAME
sam_import \- Imports media to the removable media robotic device
.SH SYNOPSIS
.LP
.BI "cc [ " "flag"
.BI " ... ] " "file"
.BI " ... -L/opt/SUNWsamfs/lib -lsamapi [ " "library" " ... ]"
.LP
.nf
.ft 3
#include "/opt/SUNWsamfs/include/samapi.h"
.ft
.fi
.LP
.BI "int sam_import(ushort_t " "eq_number" ,
.BI "char *" "vsn" ,
.BI "char *" "media_nm" ,
.BI "int " "audit_eod" ,
.BI "int " "wait_response" );
.SH DESCRIPTION
.PP
.B sam_import(\|)
requests the robotic media changer to import the selected media into the
robot.  The media is placed in
the first available element address of the catalog.
For GRAU, STK, or IBM libraries, the
.I vsn
must be specified.  Physical import of media within the GRAU and STK are
performed by utilites supplied by the vendor.  Both the
.I vsn
and the
.I media_nm
need to be specified to import into the Historian.  
.PP
The call will return immediately after issuing the command if
zero is specified for
.I wait_response
value.  Other values for
.I wait_response
will give undefined results.
.SH "RETURN VALUES"
Upon succesful completion a value of 0 is returned.
Otherwise, a value of \-1 is returned and
\f4errno\fP
is set to indicate the error.
.SH ERRORS
.PP
.B sam_import(\|)
fails if one or more of the following are true:
.TP 20
.SB ER_AUDIT_EOD_NOT_HISTORIAN
A non-zero
.I audit_eod
flag, requesting an audit to end-of-data, cannot be specified for the
Historian.
.TP
.SB ER_DEVICE_NOT_READY
The specified
.I eq_number
device is not ready.
.TP
.SB ER_FIFO_PATH_LENGTH
The path and filename for the FIFO pipe is too long.
.TP
.SB ER_HISTORIAN_MEDIA_ONLY
.I media_nm
can only be specified for the Historian.
.TP
.SB ER_INVALID_MEDIA_TYPE
The specified
.I media_nm
is not a valid media type.
.TP
.SB ER_MEDIA_FOR_HISTORIAN
The type of media
.I media_nm
must be specified for the Historian.
.TP
.SB ER_NO_EQUIP_ORDINAL
.I eq_number
is not a defined equipment number in this configuration.
.TP
.SB ER_NO_MASTER_SHM
No Sun StorEdge \%SAM-FS or Sun \%SAM-QFS master shared memory segment defined.
Check that the Sun StorEdge \%SAM-FS or Sun \%SAM-QFS file systems are mounted.
.TP
.SB ER_NO_RESPONSE_FIFO
Unable to create the response FIFO pipe.
.TP
.SB ER_NO_MASTER_SHM_ATT
No Sun StorEdge \%SAM-FS or Sun \%SAM-QFS master shared memory segment found.
Check that the Sun StorEdge \%SAM-FS or Sun \%SAM-QFS file systems are mounted.
.TP
.SB ER_OPERATOR_NOT_PRIV
Operator does not have permission to import removable media.
.TP
.SB ER_ROBOT_DEVICE_REQUIRED
Equipment number
.I eq_number
is not defined as a robotic device.
.TP
.SB ER_VSN_BARCODE_REQUIRED
.I vsn
must be specified for the GRAU, STK, and IBM media changers and the Historian.
.SH FILES
.TP 20
.SB mcf
The configuration file for Sun StorEdge \%SAM-FS and Sun \%SAM-QFS file systems.
.SH SEE ALSO
.BR export (8),
.BR import (8),
.BR sam-robotsd (8).
.PP
.BR sam_export (3).
.PP
.BR mcf (5).


src/lib/samapi/sam_import.c
/*
 * sam_import.c - import removable media API
 *
 *	sam_import() causes the media specified to be input into the removable
 *	media device.
 *
 */


/*
 *    SAM-QFS_notice_begin
 *
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
 * or https://illumos.org/license/CDDL.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at pkg/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 *
 *    SAM-QFS_notice_end
 */

#pragma ident "$Revision: 1.18 $"


#include <stdio.h>
#include <fcntl.h>
#include <string.h>
#include <errno.h>
#include <sys/param.h>
#include <sys/types.h>

#include "sam/types.h"
#include "sam/param.h"
#include "aml/device.h"
#include "aml/fifo.h"
#include "sam/lib.h"
#include "aml/proto.h"
#include "sam/devnm.h"
#include "aml/samapi.h"
#include "aml/sam_utils.h"
#include "pub/devstat.h"
#include "pub/sam_errno.h"




/*
 *	sam_import() - API function to import media
 *
 *	Input parameters --
 *		eq_ord		Equipment ordinal
 *		vsn		Volume serial number to import or(char *) NULL
 *		media_type	Type of media to import(Historian)
 *				or(char *) NULL
 *		audit_eod	Flag to indicate that audit to EOD performed
 *		wait_response	Wait for command response if nonzero
 *				-1	Wait forever
 *				> zero	Wait this many seconds for response
 *
 *	Output parameters --
 *		errno		Set to error number if error encountered
 *
 *	Return value --
 *		0		If successful in exporting the media
 *		-1		If error encountered,
 *				'errno' set to error number
 *
 */

int
sam_import(
	ushort_t eq_ord,	/* Equipment ordinal */
	char	*vsn,		/* Volume Serial Number */
	char	*media_nm,	/* Media nmemonic string */
	int	audit_eod,	/* Flag for audit to EOD performed */
	int	wait_response)	/* Set nonzero value to wait for response */
{
	char *fifo_path;	/* Path to FIFO pipe, from shared memory */
	int media = 0;
	dev_ent_t *device;
	operator_t operator;	/* Data on operator */
	sam_cmd_fifo_t cmd_block;


	/*
	 * Convert media nmemonic to device type if needed
	 */

	if ((char *)NULL != media_nm) {
		if ((media = media_to_device(media_nm)) == -1) {
			errno = ER_INVALID_MEDIA_TYPE;
			return (-1);
		}
	}

	/*
	 * Get device entry and FIFO path from master shared memory segment
	 */

	if (sam_get_dev(eq_ord, &device, &fifo_path, &operator) < 0) {
		return (-1);
	}

	/*
	 * If user cannot perform movement of media within a library, return
	 */

	if (!SAM_OPER_SLOT(operator)) {
		errno = ER_OPERATOR_NOT_PRIV;
		return (-1);
	}

	/*
	 * If device is not ready for import, return
	 */

	if (!device->status.b.ready) {
		errno = ER_DEVICE_NOT_READY;
		return (-1);
	}

	/*
	 * If device is not a robot, return
	 */

	if (!IS_ROBOT(device)) {
		errno = ER_ROBOT_DEVICE_REQUIRED;
		return (-1);
	}

	/*
	 * If device is not the Historian and media type specified, return
	 */

	if ((device->equ_type != DT_HISTORIAN) && (media != 0)) {
		errno = ER_HISTORIAN_MEDIA_ONLY;
		return (-1);
	}

	/*
	 * If device is the Historian and audit to end of data specified, return
	 */

	if ((device->equ_type == DT_HISTORIAN) && audit_eod) {
		errno = ER_AUDIT_EOD_NOT_HISTORIAN;
		return (-1);
	}

	/*
	 * If device is the Historian and no media nor VSN specified, return
	 */

	if (device->equ_type == DT_HISTORIAN) {
		if (media == 0) {
			errno = ER_MEDIA_FOR_HISTORIAN;
			return (-1);
		}

		if ((char *)NULL == vsn) {
			errno = ER_VSN_BARCODE_REQUIRED;
			return (-1);
		}

		cmd_block.cmd = CMD_FIFO_ADD_VSN;
		cmd_block.media = media;
	}
	else
	{
		/*
		 * If device is a robot and no VSN is specified for a GRAU,
		 * STK, or IBM library, return with error
		 */

		if (((char *)NULL == vsn) && ((DT_GRAUACI == device->type) ||
		    (DT_STKAPI == device->type) ||
		    (DT_IBMATL == device->type) ||
		    (DT_SONYPSC == device->type))) {
			errno = ER_VSN_BARCODE_REQUIRED;
			return (-1);
		}
		cmd_block.cmd = CMD_FIFO_IMPORT;

		if (audit_eod) {
			cmd_block.flags |= CMD_IMPORT_AUDIT;
		}
	}

	cmd_block.magic = CMD_FIFO_MAGIC;
	cmd_block.slot = ROBOT_NO_SLOT;
	cmd_block.eq = eq_ord;

	if ((char *)NULL != vsn) {
		strncpy(cmd_block.vsn, vsn, sizeof (vsn_t));
	}

	if (sam_send_cmd(&cmd_block, wait_response, fifo_path) < 0) {
		return (-1);
	}

	return (0);
}
